dbms/src/Flash/Mpp/MPPTunnel.cpp

```diff
@@ -176,196 +167,229 @@ void MPPTunnelBase<Writer>::write(const mpp::MPPDataPacket & data, bool close_af
+    LOG_FMT_TRACE(log, "ready to finish, is_local: {}", mode == TunnelSenderMode::LOCAL);
     {
         std::unique_lock lk(mu);
-        if (finished)
-            throw Exception("write to tunnel which is already closed," + consumer_state.getError());
+        if (status == TunnelStatus::Finished)
+            throw Exception("write to tunnel which is already closed," + tunnel_sender->getConsumerFinishMsg());
         /// make sure to finish the tunnel after it is connected
         waitUntilConnectedOrFinished(lk);
         finishSendQueue();
     }
-    waitForConsumerFinish(/*allow_throw=*/true);
-}
-
-template <typename Writer>
-std::shared_ptr<mpp::MPPDataPacket> MPPTunnelBase<Writer>::readForLocal()
-{
-    RUNTIME_ASSERT(is_local, log, "should not reach readForLocal for remote tunnels");
-    MPPDataPacketPtr res;
-    if (send_queue.pop(res))
-        return res;
-    consumerFinish("");
-    return nullptr;
+    waitForSenderFinish(/*allow_throw=*/true);
 }
 
-template <typename Writer>
-void MPPTunnelBase<Writer>::connect(Writer * writer_)
+void MPPTunnel::connect(PacketWriter * writer)
 {
     {
         std::unique_lock lk(mu);
-        if (connected)
-            throw Exception("MPPTunnel has connected");
-        if (finished)
-            throw Exception("MPPTunnel has finished");
+        if (status != TunnelStatus::Unconnected)
+            throw Exception(fmt::format("MPPTunnel has connected or finished: {}", statusToString()));
 
         LOG_FMT_TRACE(log, "ready to connect");
-        if (is_local)
-            RUNTIME_ASSERT(writer_ == nullptr, log);
-        else
+        switch (mode)
         {
-            writer = writer_;
-            if (!is_async)
-            {
-                // communicate send_thread through `consumer_state`
-                // NOTE: if the thread creation failed, `connected` will still be `false`.
-                thread_manager->schedule(true, "MPPTunnel", [this] {
-                    sendJob();
-                });
-            }
+        case TunnelSenderMode::LOCAL:
+            RUNTIME_ASSERT(writer == nullptr, log);
+            local_tunnel_sender = std::make_shared<LocalTunnelSender>(mode, send_queue, nullptr, log, tunnel_id);
+            tunnel_sender = local_tunnel_sender;
+            break;
+        case TunnelSenderMode::SYNC_GRPC:
+            sync_tunnel_sender = std::make_shared<SyncTunnelSender>(mode, send_queue, writer, log, tunnel_id);
+            sync_tunnel_sender->startSendThread();
+            tunnel_sender = sync_tunnel_sender;
+            break;
+        case TunnelSenderMode::ASYNC_GRPC:
+            async_tunnel_sender = std::make_shared<AsyncTunnelSender>(mode, send_queue, writer, log, tunnel_id);
+            tunnel_sender = async_tunnel_sender;
+            RUNTIME_ASSERT(writer != nullptr, log, "Async writer shouldn't be null");
+            RUNTIME_ASSERT(tunnel_sender != nullptr, log, "Tunnel sender shouldn't be null");
```

Comment:
Why add this assert, line 210 already make sure that `tunnel_sender` is not nullptr

Reply:
You're right. I'll remove the check.